ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs

```diff
@@ -803,15 +855,26 @@ withPeerStateActions timeout
           throwIO (MiniProtocolExceptions errs)
 
         Just AllSucceeded -> do
-          atomically $ do
-            writeTVar pchPeerState (PeerStatus PeerWarm)
-            -- We need to update hot protocols to indicate that they are not
-            -- running.
-            stateTVar (getMiniProtocolsVar TokHot pchAppHandles)
-                      (\a -> ( Map.map (const (pure NotRunning)) a
-                             , ()
-                             ))
-          traceWith spsTracer (PeerStatusChanged (HotToWarm pchConnectionId))
+          suc <- atomically $ do
+            -- Only set the status to PeerWarm if the peer isn't PeerCold
+            -- (can happen asynchronously).
+            notCold <- updateUnlessCold pchPeerState (PeerStatus PeerWarm)
+            when notCold $ do
+              -- We need to update hot protocols to indicate that they are not
+              -- running.
+              stateTVar (getMiniProtocolsVar TokHot pchAppHandles)
+                        (\a -> ( Map.map (const (pure NotRunning)) a
+                               , ()
+                               ))
```

Comment:
Hmm, why this even compiles? `stateTVar :: TVar s -> (s -> (a, s)) -> STM a`, the callback returns `( _, ())` as if the state was `s ~ ()`.  Or does `stateTVar` from `io-sim-classes` has a different type signature?

Reply:
Yes the code compiles. I have only indented the code, it is exactly as it was before my patch.

Reply:
I found the problem:
```
Control.Class.MonadSTM.Strict.stateTVar :: StrictTVar m s -> (s -> (s, a)) -> STM m a
```
and
```
Control.Class.MonadSTM.stateTVar :: TVar m s -> (s -> (a, s)) -> STM m a
```
The latter one is the right signature.